toolz/functoolz.py

```diff
@@ -67,17 +106,30 @@ def thread_first(val, *forms):
     See Also:
         thread_last
     """
-    def evalform_front(val, form):
+
+    @overload
+    def evalform_front(val: _S, form: TransformOp[_S, _T]) -> _T: ...
+
+    @overload
+    def evalform_front(
+        val: _S, form: TupleTransformFront[_S, Unpack[_Ts], _T]
+    ) -> _T: ...
+
+    def evalform_front(
+        val: _S,
+        form: TransformOp[_S, _T] | TupleTransformFront[_S, Unpack[_Ts], _T],
+    ) -> _T:
         if callable(form):
             return form(val)
-        if isinstance(form, tuple):
-            func, args = form[0], form[1:]
-            args = (val,) + args
-            return func(*args)
-    return reduce(evalform_front, forms, val)
+        # if isinstance(form, tuple):
```

Comment:
according to the types, no need to check for `tuple` type, but it would be good to check anyway.